File: Autonomous Mobile Robots Course/Chapter9/Lesson2/Chapter9_Lesson2.m
% Chapter9_Lesson2.m
% Autonomous Mobile Robots — Chapter 9, Lesson 2
% Log-Odds Updates and Bayesian Cells (Occupancy Grid Mapping)
%
% This script demonstrates a log-odds occupancy grid update using a simple
% inverse sensor model and Bresenham ray traversal.
%
% Notes:
%   - Requires base MATLAB for the script portion.
%   - Optional: Simulink model construction is included at the end (requires Simulink).

clear; clc; close all;

%% Parameters
W = 220; H = 220;
res = 0.05;            % meters per cell
origin = [-5.5, -5.5]; % world coordinate of cell (1,1) center
rangeMax = 6.0;

p0 = 0.5;
pOcc = 0.7;
pFree = 0.3;
lMin = -10; lMax = 10;

logOdds = zeros(H, W, 'single');

%% Helpers
logit = @(p) log(max(min(p, 1-1e-6), 1e-6) ./ (1 - max(min(p, 1-1e-6), 1e-6)));
sigmoid = @(l) 1 ./ (1 + exp(-l));

worldToMap = @(x,y) [round((x - origin(1)) / res) + 1, round((y - origin(2)) / res) + 1]; % 1-based

inBounds = @(ix,iy) (ix >= 1 && ix <= W && iy >= 1 && iy <= H);

updateCell = @(ix,iy,dl) assignin('caller','logOdds', ...
  localUpdate(logOdds, ix, iy, dl, lMin, lMax, inBounds));

%% Synthetic environment: two circles
circles = [ 1.5, 0.5, 0.6;
           -1.2, 1.2, 0.5];

poses = [ -2.0, -2.0, 0.2;
           0.0, -2.5, 0.6;
           2.0, -1.5, 1.0;
           2.0,  1.0, 1.6;
           0.0,  2.5, 2.5;
          -2.0,  2.0,-2.7];

angles = linspace(-pi/2, pi/2, 121);

%% Main loop: update scans
l0 = logit(p0);
lOcc = logit(pOcc) - l0;
lFree = logit(pFree) - l0;

for k = 1:size(poses,1)
    x = poses(k,1); y = poses(k,2); th = poses(k,3);

    ranges = zeros(size(angles));
    for i = 1:numel(angles)
        ranges(i) = raycastCircles(x, y, th, angles(i), rangeMax, circles);
    end

    s = worldToMap(x,y);
    sx = s(1); sy = s(2);

    for i = 1:numel(angles)
        a = angles(i); r = ranges(i);
        if ~isfinite(r) || r <= 0, continue; end
        rEff = min(r, rangeMax);

        bx = x + rEff * cos(th + a);
        by = y + rEff * sin(th + a);
        e = worldToMap(bx,by);
        ex = e(1); ey = e(2);

        ray = bresenhamLine(sx, sy, ex, ey);
        if isempty(ray), continue; end

        % Free cells (excluding endpoint)
        for j = 1:size(ray,1)-1
            ix = ray(j,1); iy = ray(j,2);
            if inBounds(ix,iy)
                logOdds(iy,ix) = min(max(logOdds(iy,ix) + lFree, lMin), lMax);
            end
        end

        % Occupied endpoint if hit
        if r < rangeMax - 1e-9
            ix = ray(end,1); iy = ray(end,2);
            if inBounds(ix,iy)
                logOdds(iy,ix) = min(max(logOdds(iy,ix) + lOcc, lMin), lMax);
            end
        end
    end
end

%% Visualize
P = sigmoid(double(logOdds));
figure('Color','w'); imagesc([origin(1), origin(1)+W*res], [origin(2), origin(2)+H*res], P);
axis image; set(gca,'YDir','normal'); colorbar;
title('Occupancy probability (log-odds fused)');
xlabel('x [m]'); ylabel('y [m]');

%% ---- Optional: programmatically build a Simulink "log-odds accumulator" block diagram ----
% This section creates a *conceptual* model that integrates delta_l updates
% and applies saturation. A full LiDAR ray-tracing sensor model is typically
% implemented in MATLAB Function blocks or external code.
%
% Requires Simulink. Uncomment to generate.
%
%{
modelName = 'Chapter9_Lesson2_LogOdds_Accumulator';
new_system(modelName); open_system(modelName);

add_block('simulink/Sources/In1', [modelName '/delta_l'], 'Position',[30 50 60 70]);
add_block('simulink/Discrete/Unit Delay', [modelName '/z^-1'], 'Position',[110 45 160 75]);
add_block('simulink/Math Operations/Add', [modelName '/add'], 'Position',[200 45 240 75]);
add_block('simulink/Discontinuities/Saturation', [modelName '/sat'], 'Position',[280 45 340 75]);
add_block('simulink/Sinks/Out1', [modelName '/l_t'], 'Position',[380 50 410 70]);

set_param([modelName '/z^-1'], 'InitialCondition', '0');
set_param([modelName '/sat'], 'LowerLimit', num2str(lMin), 'UpperLimit', num2str(lMax));

add_line(modelName, 'delta_l/1', 'add/2');
add_line(modelName, 'z^-1/1', 'add/1');
add_line(modelName, 'add/1', 'sat/1');
add_line(modelName, 'sat/1', 'l_t/1');
add_line(modelName, 'sat/1', 'z^-1/1');

save_system(modelName);
%}

%% ---- Local functions ----
function logOddsOut = localUpdate(logOddsIn, ix, iy, dl, lMin, lMax, inBoundsFun)
    logOddsOut = logOddsIn;
    if inBoundsFun(ix,iy)
        logOddsOut(iy,ix) = min(max(logOddsOut(iy,ix) + dl, lMin), lMax);
    end
end

function r = raycastCircles(x,y,th,a,rangeMax,circles)
    dx = cos(th+a); dy = sin(th+a);
    step = 0.02;
    t = 0.0;
    while t <= rangeMax
        px = x + t*dx; py = y + t*dy;
        for i = 1:size(circles,1)
            cx = circles(i,1); cy = circles(i,2); rr = circles(i,3);
            if (px-cx)^2 + (py-cy)^2 <= rr^2
                r = max(0.05, t);
                return;
            end
        end
        t = t + step;
    end
    r = rangeMax;
end

function pts = bresenhamLine(x0,y0,x1,y1)
    dx = abs(x1-x0);
    dy = abs(y1-y0);
    sx = 1; if x1 < x0, sx = -1; end
    sy = 1; if y1 < y0, sy = -1; end

    x = x0; y = y0;
    pts = [];

    if dy <= dx
        err = floor(dx/2);
        while x ~= x1
            pts = [pts; x, y]; %#ok<AGROW>
            err = err - dy;
            if err < 0
                y = y + sy;
                err = err + dx;
            end
            x = x + sx;
        end
        pts = [pts; x1, y1];
    else
        err = floor(dy/2);
        while y ~= y1
            pts = [pts; x, y]; %#ok<AGROW>
            err = err - dx;
            if err < 0
                x = x + sx;
                err = err + dy;
            end
            y = y + sy;
        end
        pts = [pts; x1, y1];
    end
end

File: Autonomous Mobile Robots Course/Chapter9/Lesson2/Chapter9_Lesson2.java
// Chapter9_Lesson2.java
// Autonomous Mobile Robots — Chapter 9, Lesson 2
// Log-Odds Updates and Bayesian Cells (Occupancy Grid Mapping)
//
// Minimal Java example that maintains a log-odds occupancy grid, updates with an inverse
// sensor model along rays (Bresenham), and writes a PGM image.
//
// Compile:
//   javac Chapter9_Lesson2.java
// Run:
//   java Chapter9_Lesson2
//
// Output: occupancy_java.pgm

import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class Chapter9_Lesson2 {

  static float clamp(float v, float lo, float hi) {
    return (v < lo) ? lo : (v > hi) ? hi : v;
  }

  static float logit(float p) {
    float eps = 1e-6f;
    p = clamp(p, eps, 1.0f - eps);
    return (float) Math.log(p / (1.0f - p));
  }

  static float logistic(float l) {
    if (l >= 0.0f) {
      float z = (float) Math.exp(-l);
      return 1.0f / (1.0f + z);
    } else {
      float z = (float) Math.exp(l);
      return z / (1.0f + z);
    }
  }

  static class Grid {
    final int width, height;
    final float res, originX, originY;
    final float[] logOdds; // row-major

    Grid(int w, int h, float res, float ox, float oy) {
      this.width = w;
      this.height = h;
      this.res = res;
      this.originX = ox;
      this.originY = oy;
      this.logOdds = new float[w * h];
    }

    boolean inBounds(int ix, int iy) {
      return (0 <= ix && ix < width && 0 <= iy && iy < height);
    }

    int idx(int ix, int iy) {
      return iy * width + ix;
    }

    int[] worldToMap(float x, float y) {
      int ix = Math.round((x - originX) / res);
      int iy = Math.round((y - originY) / res);
      return new int[] {ix, iy};
    }

    void updateCell(int ix, int iy, float delta, float lmin, float lmax) {
      if (!inBounds(ix, iy)) return;
      int k = idx(ix, iy);
      logOdds[k] = clamp(logOdds[k] + delta, lmin, lmax);
    }
  }

  static List<int[]> bresenham(int x0, int y0, int x1, int y1) {
    List<int[]> pts = new ArrayList<>();
    int dx = Math.abs(x1 - x0);
    int dy = Math.abs(y1 - y0);
    int sx = (x1 >= x0) ? 1 : -1;
    int sy = (y1 >= y0) ? 1 : -1;
    int x = x0, y = y0;

    if (dy <= dx) {
      int err = dx / 2;
      while (x != x1) {
        pts.add(new int[] {x, y});
        err -= dy;
        if (err < 0) {
          y += sy;
          err += dx;
        }
        x += sx;
      }
      pts.add(new int[] {x1, y1});
    } else {
      int err = dy / 2;
      while (y != y1) {
        pts.add(new int[] {x, y});
        err -= dx;
        if (err < 0) {
          x += sx;
          err += dy;
        }
        y += sy;
      }
      pts.add(new int[] {x1, y1});
    }
    return pts;
  }

  static float raycastCircles(float x, float y, float th, float a, float rangeMax, float[][] circles) {
    float dx = (float) Math.cos(th + a);
    float dy = (float) Math.sin(th + a);
    float step = 0.02f;
    for (float t = 0.0f; t <= rangeMax; t += step) {
      float px = x + t * dx;
      float py = y + t * dy;
      for (float[] c : circles) {
        float cx = c[0], cy = c[1], rr = c[2];
        float ex = px - cx, ey = py - cy;
        if (ex * ex + ey * ey <= rr * rr) {
          return Math.max(0.05f, t);
        }
      }
    }
    return rangeMax;
  }

  static void updateScan(Grid g, float x, float y, float th,
                         float[] angles, float[] ranges, float rangeMax,
                         float p0, float pOcc, float pFree,
                         float lmin, float lmax) {
    float l0 = logit(p0);
    float lOcc = logit(pOcc) - l0;
    float lFree = logit(pFree) - l0;

    int[] s = g.worldToMap(x, y);
    int sx = s[0], sy = s[1];

    for (int i = 0; i < angles.length; i++) {
      float a = angles[i];
      float r = ranges[i];
      if (!Float.isFinite(r) || r <= 0.0f) continue;
      float rEff = Math.min(r, rangeMax);

      float bx = x + rEff * (float) Math.cos(th + a);
      float by = y + rEff * (float) Math.sin(th + a);
      int[] e = g.worldToMap(bx, by);
      int ex = e[0], ey = e[1];

      List<int[]> ray = bresenham(sx, sy, ex, ey);
      if (ray.isEmpty()) continue;

      for (int k = 0; k + 1 < ray.size(); k++) {
        int[] p = ray.get(k);
        g.updateCell(p[0], p[1], lFree, lmin, lmax);
      }
      if (r < rangeMax - 1e-6f) {
        int[] p = ray.get(ray.size() - 1);
        g.updateCell(p[0], p[1], lOcc, lmin, lmax);
      }
    }
  }

  static void savePGM(Grid g, String path) throws IOException {
    try (FileOutputStream out = new FileOutputStream(path)) {
      String header = "P5\n" + g.width + " " + g.height + "\n255\n";
      out.write(header.getBytes(StandardCharsets.US_ASCII));

      for (int iy = g.height - 1; iy >= 0; iy--) {
        for (int ix = 0; ix < g.width; ix++) {
          float l = g.logOdds[g.idx(ix, iy)];
          float p = logistic(l);
          int v = Math.round(clamp(255.0f * (1.0f - p), 0.0f, 255.0f));
          out.write((byte) (v & 0xFF));
        }
      }
    }
  }

  public static void main(String[] args) throws IOException {
    Grid grid = new Grid(220, 220, 0.05f, -5.5f, -5.5f);

    float[][] circles = new float[][] {
      {1.5f, 0.5f, 0.6f},
      {-1.2f, 1.2f, 0.5f},
    };

    // Angles
    int N = 121;
    float[] angles = new float[N];
    for (int i = 0; i < N; i++) {
      angles[i] = (float) (-Math.PI / 2.0 + (double) i * (Math.PI / (N - 1)));
    }
    float rangeMax = 6.0f;

    float[][] poses = new float[][] {
      {-2.0f, -2.0f, 0.2f},
      {0.0f, -2.5f, 0.6f},
      {2.0f, -1.5f, 1.0f},
      {2.0f, 1.0f, 1.6f},
      {0.0f, 2.5f, 2.5f},
      {-2.0f, 2.0f, -2.7f},
    };

    for (float[] pose : poses) {
      float x = pose[0], y = pose[1], th = pose[2];
      float[] ranges = new float[N];
      for (int i = 0; i < N; i++) {
        ranges[i] = raycastCircles(x, y, th, angles[i], rangeMax, circles);
      }
      updateScan(grid, x, y, th, angles, ranges, rangeMax,
                 0.5f, 0.7f, 0.3f, -10.0f, 10.0f);
    }

    savePGM(grid, "occupancy_java.pgm");
    System.out.println("Wrote occupancy_java.pgm");
  }
}
